# Actividad 5.1 Matriz HAT

Samuel García Berenfeld - A01642317

Israel Vidal Paredes - A01750543

#### 1. Para el siguiente conjunto de datos, obtener la Matriz HAT, H, los valores predichos, y^ y los coeficientes de la ecuación de regresión β.

**(Proceso al final del documento pdf)**



####2. Verifiar los resultados utilizando Python

In [ ]:
import numpy as np

X = np.array([[1, 2],
              [1, 3],
              [1, 5],
              [1, 7],
              [1, 9]])

y = np.array([5, 8, 7, 10, 12])

XtX_inv = np.linalg.inv(X.T @ X)
beta = XtX_inv @ X.T @ y

y_hat = X @ beta

H = X @ XtX_inv @ X.T

print("Coeficientes de regresión beta:")
print(beta)
print("\nValores predichos y^:")
print(y_hat)
print("\nMatriz HAT H:")
print(H)


Coeficientes de regresión beta:
[3.86585366 0.87195122]

Valores predichos y^:
[ 5.6097561   6.48170732  8.22560976  9.9695122  11.71341463]

Matriz HAT H:
[[ 0.51219512  0.41463415  0.2195122   0.02439024 -0.17073171]
 [ 0.41463415  0.34756098  0.21341463  0.07926829 -0.05487805]
 [ 0.2195122   0.21341463  0.20121951  0.18902439  0.17682927]
 [ 0.02439024  0.07926829  0.18902439  0.29878049  0.40853659]
 [-0.17073171 -0.05487805  0.17682927  0.40853659  0.6402439 ]]


####3. Utilizando los Datos “Cirugía de Hígado” obtener la matriz HAT, los valores predichos, y, y los coeficientes de la ecuación de regresión β utilizando el método de matrices. (Puede realizarse con cualquier paquete).

In [ ]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import MinMaxScaler

file_path = '/content/sample_data/Covarianza, Correlación y HAT.csv'
data = pd.read_csv(file_path, encoding='latin1')

X = data[['Factor Coagulación', 'Índice pronóstico', 'Función de enzima',
          'Función de hígado', 'Edad', 'Género',
          'Alcohol (moderado)', 'Alcohol (severo)']]

y = data['Sobrevivencia (días)']

scaler = MinMaxScaler()
X_scaled = scaler.fit_transform(X)

X_scaled = np.hstack([np.ones((X_scaled.shape[0], 1)), X_scaled])

XtX_inv = np.linalg.inv(X_scaled.T @ X_scaled)
beta = XtX_inv @ X_scaled.T @ y

y_hat = X_scaled @ beta

H = X_scaled @ XtX_inv @ X_scaled.T

coef_df = pd.DataFrame(beta, index=['Intercepto'] + list(data.columns[:-1]), columns=['Coeficientes'])

y_hat_df = pd.DataFrame(y_hat, columns=['Valores Predichos'])

H_df = pd.DataFrame(H, columns=[f'Col_{i+1}' for i in range(H.shape[1])])
H_df_limited = H_df.iloc[:20, :20]

print("Coeficientes de regresión:")
print(coef_df.round(4))

print("\nValores predichos:")
print(y_hat_df.round(4))

print("\nMatriz HAT (primeros 20x20):")
print(H_df_limited.round(4))


Coeficientes de regresión:
                    Coeficientes
Intercepto             -575.8637
Factor Coagulación      453.2609
Índice pronóstico       738.2324
Función de enzima       854.1430
Función de hígado       429.4316
Edad                     25.6389
Género                   13.0926
Alcohol (moderado)      -41.2676
Alcohol (severo)        195.7070

Valores predichos:
     Valores Predichos
0             706.2562
1             430.8229
2             732.2298
3             425.0396
4            1454.5876
5             317.9851
6             561.4343
7             620.5703
8             814.9924
9             761.6839
10            989.9519
11            228.0218
12           1394.1346
13           1050.2917
14            937.9567
15            732.4843
16            418.0998
17             58.3229
18            857.9967
19            939.0347
20            555.5315
21            366.0251
22            485.5213
23            817.7782
24            861.3188
25            669.3581
26

#### 4. Concluir sobre el significado de los valores de apalancamiento y formular la ecuación de regresión.

En general, al utilizar el método de la matriz hat para crear una regresión, podemos obtener cada valor beta de beta 0 hasta beta n, el primero representando el intercepto de la función y los demás el coeficiente de cada una de las variables que utilizamos para la predicción. Después, se puede utilizar esta ecuacuión para predecir los valores de nuestra variable de respuesta.

Por otro lado, la matriz HAT en sí consiste en un conjunto de valores conocidos como valores de apalancamiento, que indican la influencia de cada una de las observaciones en la tabla de datos original. Al analizar estos valores, podemos identificar si algún outlier tiene una influencia desproporcionada en el modelo de regresión. Si se detecta una influencia alta de un outlier, podríamos considerar no incluir dicha observación en futuras predicciones para evitar sesgos en el modelo.

Una manera en la que esto se puede decidir, es obteniendo el apalancamiento promedio, que es el número de predictores usados entre el numero de observaciones, y si al comparar algún valor en la diagonal de la matriz HAT un valor sea mayor a este promedio, sería una indicación fuerte que se cuenta con un outlier.

La ecuación de regresión final es:


ŷ = -575.8637 + 453.2609 * (Factor Coagulación) + 738.2324 * (Índice pronóstico) + 854.1430 * (Función de enzima) + 429.4316 * (Función de hígado) + 25.6389  * (Edad) + 13.0926  * (Género) - 41.2676  * (Alcohol moderado) + 195.7070 * (Alcohol severo)

